### PR TITLE
8353845: com.sun.javafx.css.BitSet.equals(null) throws NPE

### PR DESCRIPTION
**Pull request by mstr2**
Fixes the bug that `BitSet.equals(null)` throws NPE.

A single reviewer should be sufficient.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8353845](https://bugs.openjdk.org/browse/JDK-8353845): com.sun.javafx.css.BitSet.equals(null) throws NPE (**Bug** - P4)

### Reviewers
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1766/head:pull/1766` \
`$ git checkout pull/1766`

Update a local copy of the PR: \
`$ git checkout pull/1766` \
`$ git pull https://git.openjdk.org/jfx.git pull/1766/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1766`

View PR using the GUI difftool: \
`$ git pr show -t 1766`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1766.diff">https://git.openjdk.org/jfx/pull/1766.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1766#issuecomment-2788568630)
</details>

### CODE REVIEW

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/css/BitSetTest.java [hunk 1]
```diff
@@ -47,6 +45,11 @@ public class BitSetTest {
     private final PseudoClass b = PseudoClass.getPseudoClass("b");
     private final PseudoClass c = PseudoClass.getPseudoClass("c");
 
+    @Test
+    void equalsNullShouldBeFalse() {
+        assertFalse(set.equals(null));
```
Comment (nlisker) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/css/BitSetTest.java:50]: Isn't `assertNotEquals` more suitable here?

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 2]
```diff
@@ -533,6 +533,11 @@ public boolean equals(Object obj) {
         if (obj == this) {
             return true;
         }
+
+        if (obj == null) {
+            return false;
+        }
+
```
Comment (arapte) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:540]: Here is a suggestion to modify this source as:

```
if (obj == null || getClass() != obj.getClass()) {
    return false;
}
return equalsBitSet((BitSet<?>) obj);
```
and with this the call to `super.equals()` won't be necessary. It seems it was unnecessary earlier too. it might have been added only to avoid missing return error.

**Review by hjohn — APPROVED** [review]
Looks like a regression I introduced when cleaning up all the bugs in this class, oops.

**Comment by andy-goryachev-oracle** [comment]
there seems to be some discussion, so let's have 2 reviewers, just to be safe.

File: modules/javafx.graphics/src/shims/java/com/sun/javafx/css/BitSetShim.java [hunk 3]
```diff
@@ -185,4 +185,8 @@ public Stream<T> stream() {
     public Stream<T> parallelStream() {
         return delegate.parallelStream();
     }
+
+    public boolean equals(Object other) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/shims/java/com/sun/javafx/css/BitSetShim.java:189]: This line generates 2 warnings in Eclipse:

```
Description	Resource	Type	Path	Location
The method equals(Object) of type BitSetShim<T> should be tagged with @Override since it actually overrides a superclass method	BitSetShim.java	Java Problem	/graphics/src/shims/java/com/sun/javafx/css	line 189
The type BitSetShim<T> should also implement hashCode() since it overrides Object.equals()	BitSetShim.java	Java Problem	/graphics/src/shims/java/com/sun/javafx/css	line 39
```

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/css/BitSetTest.java [hunk 4]
```diff
@@ -47,6 +45,11 @@ public class BitSetTest {
...
+    }
+
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/css/BitSetTest.java:52]: Since we are here, should we add a case to test `equals()` with a non-empty `HashSet` or `Set.of()`, similar to `twoEmptyBitSetsShouldBeEqual()`?

**Review by andy-goryachev-oracle — CHANGES_REQUESTED** [review]

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 5]
```diff
@@ -533,6 +533,11 @@ public boolean equals(Object obj) {
         if (obj == this) {
             return true;
         }
+
+        if (obj == null) {
```
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:537]: That's unnecessary nesting, which should be avoided at all times IMHO.

**Review by andy-goryachev-oracle — APPROVED** [review]
thank you for fixing the bug and making the changes!

**Comment by andy-goryachev-oracle** [comment]
https://github.com/openjdk/jfx/pull/1766/checks?check_run_id=40273913986

this is a weird failure, I've never seen it before.  something happened at cygwin (after yesterday outage) perhaps?

@mstr2 could you re-run the windows job please?

**Review by hjohn — APPROVED** [review]

**Comment by mstr2** [comment]
> https://github.com/openjdk/jfx/pull/1766/checks?check_run_id=40273913986
> 
> this is a weird failure, I've never seen it before. something happened at cygwin (after yesterday outage) perhaps?
> 
> @mstr2 could you re-run the windows job please?

Seems like the failure is persistent.

**Comment by kevinrushforth** [comment]
> > https://github.com/openjdk/jfx/pull/1766/checks?check_run_id=40273913986
> > this is a weird failure, I've never seen it before. something happened at cygwin (after yesterday outage) perhaps?
> > @mstr2 could you re-run the windows job please?
> 
> Seems like the failure is persistent.

That's what I see, too. I pushed the current master branch to a new `test-build` branch in my repo to trigger a GHA build. Same error:

https://github.com/kevinrushforth/jfx/actions/runs/14364598377/job/40274343597

I'll look into this and file a follow-up bug later, but it's unrelated to this PR nor does not impact its readiness to integrate.
